toolchain/check/testdata/class/no_prelude/destroy.carbon

```diff
@@ -8,161 +8,430 @@
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  fn destroy[self: Self];
+}
 
-// --- fail_todo_missing_params.carbon
+// --- fail_explicit_params.carbon
 
 library "[[@TEST_NAME]]";
 
+class C {
+  // CHECK:STDERR: fail_explicit_params.carbon:[[@LINE+4]]:26: error: unexpected parameter [DestroyFunctionNonEmptyExplicitParams]
+  // CHECK:STDERR:   fn destroy[self: Self](x: ());
+  // CHECK:STDERR:                          ^~~~~
+  // CHECK:STDERR:
+  fn destroy[self: Self](x: ());
+}
+
+// --- fail_return_type.carbon
+
+library "[[@TEST_NAME]]";
 
 class C {
-  // CHECK:STDERR: fail_todo_missing_params.carbon:[[@LINE+4]]:6: error: semantics TODO: `KeywordNameNotBeforeParamsId` [SemanticsTodo]
-  // CHECK:STDERR:   fn destroy;
-  // CHECK:STDERR:      ^~~~~~~
+  // CHECK:STDERR: fail_return_type.carbon:[[@LINE+4]]:28: error: incorrect return type; must be unspecified or `()` [DestroyFunctionIncorrectReturnType]
+  // CHECK:STDERR:   fn destroy[self: Self]() -> {};
+  // CHECK:STDERR:                            ^~~~~
   // CHECK:STDERR:
-  fn destroy;
+  fn destroy[self: Self]() -> {};
+}
+
+// --- fail_out_of_line_missing_params.carbon
+
+library "[[@TEST_NAME]]";
+
+
+class C {
+  fn destroy[self: Self]();
 }
 
+// CHECK:STDERR: fail_out_of_line_missing_params.carbon:[[@LINE+7]]:1: error: redeclaration differs because of missing implicit parameter list [RedeclParamListDiffers]
+// CHECK:STDERR: fn C.destroy {}
+// CHECK:STDERR: ^~~~~~~~~~~~~~
+// CHECK:STDERR: fail_out_of_line_missing_params.carbon:[[@LINE-6]]:3: note: previously declared with implicit parameter list [RedeclParamListPrevious]
+// CHECK:STDERR:   fn destroy[self: Self]();
+// CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~~~
+// CHECK:STDERR:
 fn C.destroy {}
 
-// --- fail_todo_wrong_scope.carbon
+// --- fail_destroy_in_file_scope.carbon
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_destroy_in_file_scope.carbon:[[@LINE+4]]:1: error: declaring `fn destroy` in non-class scope [DestroyFunctionOutsideClass]
+// CHECK:STDERR: fn destroy[self: ()]();
+// CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~
+// CHECK:STDERR:
+fn destroy[self: ()]();
+
+// --- fail_destroy_in_namespace_scope.carbon
+
+library "[[@TEST_NAME]]";
+
+namespace NS;
+
+// CHECK:STDERR: fail_destroy_in_namespace_scope.carbon:[[@LINE+4]]:1: error: declaring `fn destroy` in non-class scope [DestroyFunctionOutsideClass]
+// CHECK:STDERR: fn NS.destroy();
+// CHECK:STDERR: ^~~~~~~~~~~~~~~~
+// CHECK:STDERR:
+fn NS.destroy();
+
+// --- fail_invalid_qualifier_with_params.carbon
 
 library "[[@TEST_NAME]]";
 
+class C {
+  fn destroy[self: Self]();
+}
 
-// CHECK:STDERR: fail_todo_wrong_scope.carbon:[[@LINE+4]]:4: error: semantics TODO: `KeywordNameBeforeParamsId` [SemanticsTodo]
-// CHECK:STDERR: fn destroy();
-// CHECK:STDERR:    ^~~~~~~
+// CHECK:STDERR: fail_invalid_qualifier_with_params.carbon:[[@LINE+7]]:6: error: name qualifiers are only allowed for entities that provide a scope [QualifiedNameInNonScope]
```

Comment:
FWIW, destroy() is a function so I would kind of expect this to parse like `.Foo()` being called on the return value (which is `()`), but the error is different since it's trying to do lookup on `destroy` itself.

Both are errors so maybe this is fine, but I wanted to point out what seems a bit surprising.

Reply:
To be sure, I don't think the diagnostic message is great. It doesn't follow style, and it might be more helpful to make the name lookup at `.Foo` the error, with `.destroy` as a note. But, I'd expect diagnostics to lean in the direction of qualified name lookup because, due to the `fn` introducer, we would treat `.Foo()` as a declaration instead of a call if it were valid.

e.g., always invalid, but only "always" because `destroy` is never scoped:

```
fn C.destroy[self: Self]().Foo() {}
```

Adding `r#` makes it potentially valid (kind of... `self: Self` probably means this still errors, but for different reasons):

```
fn C.r#destroy[self: Self]().Foo() {}
```

For example:

```
class C {
  class r#destroy[self: Self]() {
    fn Foo();
  }
}
```

Reply:
Ah, yeah that's a good point about the `fn` introducer, I wasn't looking at that part.